generator/internal/stub_rest_generator.cc

```diff
@@ -0,0 +1,215 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "generator/internal/stub_rest_generator.h"
+#include "absl/memory/memory.h"
+#include "absl/strings/str_split.h"
+#include "generator/internal/codegen_utils.h"
+#include "generator/internal/predicate_utils.h"
+#include "generator/internal/printer.h"
+#include <google/protobuf/descriptor.h>
+
+namespace google {
+namespace cloud {
+namespace generator_internal {
+
+StubRestGenerator::StubRestGenerator(
+    google::protobuf::ServiceDescriptor const* service_descriptor,
+    VarsDictionary service_vars,
+    std::map<std::string, VarsDictionary> service_method_vars,
+    google::protobuf::compiler::GeneratorContext* context)
+    : ServiceCodeGenerator("stub_rest_header_path", "stub_rest_cc_path",
+                           service_descriptor, std::move(service_vars),
+                           std::move(service_method_vars), context) {}
+
+Status StubRestGenerator::GenerateHeader() {
+  HeaderPrint(CopyrightLicenseFileHeader());
+  HeaderPrint(  // clang-format off
+    "\n"
+    "// Generated by the Codegen C++ plugin.\n"
+    "// If you make any local changes, they will be lost.\n"
+    "// source: $proto_file_name$\n"
+    "\n"
+    "#ifndef $header_include_guard$\n"
+    "#define $header_include_guard$\n");
+  // clang-format on
+
+  HeaderPrint("\n");
+  HeaderLocalIncludes({"google/cloud/internal/rest_client.h",
+                       "google/cloud/internal/rest_context.h",
+                       "google/cloud/status_or.h", "google/cloud/version.h"});
+  std::vector<std::string> additional_pb_header_paths =
+      absl::StrSplit(vars("additional_pb_header_paths"), absl::ByChar(','));
+  HeaderSystemIncludes(additional_pb_header_paths);
+  HeaderSystemIncludes(
+      {vars("proto_header_path"),
+       HasLongrunningMethod() ? "google/longrunning/operations.pb.h" : "",
+       "memory"});
+
+  auto result = HeaderOpenNamespaces(NamespaceType::kInternal);
+  if (!result.ok()) return result;
+
+  // Abstract interface Stub base class
+  HeaderPrint(  // clang-format off
+    "\n"
+    "class $stub_rest_class_name$ {\n"
+    " public:\n"
+    "  virtual ~$stub_rest_class_name$() = default;\n");
+  // clang-format on
+
+  for (auto const& method : methods()) {
+    HeaderPrintMethod(
+        method,
+        {MethodPattern(
```

Comment:
I am going to insist that we stop adding these predicate-based functions and we use raw strings.  The code for things like the round robin decorator is much simpler to grok:

https://github.com/googleapis/google-cloud-cpp/blob/main/generator/internal/round_robin_decorator_generator.cc